packages/framer-motion/src/value/types/complex/index.ts

```diff
@@ -16,50 +18,99 @@ function test(v: any) {
     )
 }
 
-export function analyseComplexValue(v: string | number) {
-    if (typeof v === "number") v = `${v}`
+export interface ComplexValueInfo {
+    value: string
+    values: Array<CSSVariableToken | Color | number>
+    numVars: number
+    numColors: number
+    numNumbers: number
+    tokenised: string
+}
 
-    const values: Array<Color | number> = []
-    let numColors = 0
-    let numNumbers = 0
+interface Tokeniser {
+    regex: RegExp
+    countKey: string
+    token: string
+    parse: (value: string) => any
+}
 
-    const colors = v.match(colorRegex)
-    if (colors) {
-        numColors = colors.length
-        // Strip colors from input so they're not picked up by number regex.
-        // There's a better way to combine these regex searches, but its beyond my regex skills
-        v = v.replace(colorRegex, colorToken)
-        values.push(...(colors.map(color.parse) as any))
-    }
+const cssVarTokeniser: Tokeniser = {
+    regex: cssVariableRegex,
+    countKey: "Vars",
+    token: "${v}",
+    parse: noop,
+}
+
+const colorTokeniser: Tokeniser = {
+    regex: colorRegex,
+    countKey: "Colors",
+    token: "${c}",
+    parse: color.parse,
+}
 
-    const numbers = v.match(floatRegex)
-    if (numbers) {
-        numNumbers = numbers.length
-        v = v.replace(floatRegex, numberToken)
-        values.push(...numbers.map(number.parse))
+const numberTokeniser: Tokeniser = {
+    regex: floatRegex,
+    countKey: "Numbers",
+    token: "${n}",
+    parse: number.parse,
+}
+
+function tokenise(
+    info: ComplexValueInfo,
+    { regex, countKey, token, parse }: Tokeniser
+) {
+    const matches = info.value.match(regex)
+
+    if (!matches) return
+
+    info["num" + countKey] = matches.length
+    info.value = info.value.replace(regex, token)
+    info.values.push(...(matches.map(parse) as any))
```

Comment:
You've prolly thought this through, but why do we rely on the index instead of updating the shape of `values`? For example, `values: {vars: [], colors: [], nums: []}`, this way, we won't need the countKey nor the counts?

Reply:
If I have thought this through I think it was six years ago and I was basically just copying what I did then. You're right. I'll take another look.

Reply:
As I'm refactoring this I remember it's so I can use a generic `mixArray` on the single values array